### ISSUE TITLE
fatal error: unexpected signal during runtime execution

### ISSUE THREAD

**Issue by mowenchuxin** [golang/go#23650]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
my go version is 1.9.3

### Does this issue reproduce with the latest release?
yes. 

### What operating system and processor architecture are you using (`go env`)?
darwin(macOS Sierra 10.12.5)/amd64

### What did you do?
I run commond "./create_android_update.h nothing 20 10", and came the errors.

[tool2.zip](https://github.com/golang/go/files/1684544/tool2.zip)

If possible, provide a recipe for reproducing the error.
I commit it. Please have a look at it. Thank U very much.

### What did you expect to see?
I hope it won't show the golang's runtime error . tips: it will runthe luajit commend after runtime error.

### What did you see instead?

`guxiaoliangdeMacBook-Pro:tool guxiaoliang$ ./create_android_update.sh nothing 20 10
11970
fatal error: unexpected signal during runtime execution
[signal 0xa code=0x2 addr=0xcd66bee pc=0xd7bb]

runtime stack:
runtime.gothrow(0xe57c8, 0x2a)
	/usr/local/go/src/runtime/panic.go:503 +0x67
runtime.sigpanic()
	/usr/local/go/src/runtime/sigpanic_unix.go:14 +0x4d
runtime.unlock(0x1438e0)
	/usr/local/go/src/runtime/lock_sema.go:110 +0x10b
incidlelocked(0xffffffff)
	/usr/local/go/src/runtime/proc.c:2783 +0x44
retake(0x368dd20a, 0x150eda0e, 0x150eda0e)
	/usr/local/go/src/runtime/proc.c:3004 +0xe8
sysmon()
	/usr/local/go/src/runtime/proc.c:2930 +0x22f
mstart()
	/usr/local/go/src/runtime/proc.c:859 +0xa5
runtime.mstart()
	/usr/local/go/src/runtime/proc.c:836 +0x80

goroutine 1 [syscall]:
syscall.Syscall(0x3, 0x10664800, 0x11d9, 0x11d9, 0x1050f630, 0x4b, 0x0)
	/usr/local/go/src/syscall/asm_darwin_386.s:20 +0x5
syscall.read(0x3, 0x10664800, 0x11d9, 0x11d9, 0x4b, 0x0, 0x0)
	/usr/local/go/src/syscall/zsyscall_darwin_386.go:970 +0x54
syscall.Read(0x3, 0x10664800, 0x11d9, 0x11d9, 0xee8a, 0x0, 0x0)
	/usr/local/go/src/syscall/syscall_unix.go:136 +0x4a
os.(*File).read(0x10362338, 0x10664800, 0x11d9, 0x11d9, 0xe1ad, 0x0, 0x0)
	/usr/local/go/src/os/file_unix.go:191 +0x70
os.(*File).Read(0x10362338, 0x10664800, 0x11d9, 0x11d9, 0x10664800, 0x0, 0x0)
	/usr/local/go/src/os/file.go:95 +0x70
bytes.(*Buffer).ReadFrom(0x20, 0x0, 0x105f5ae0, 0x0, 0x105f5ae0, 0x17a0fc, 0x0)
	/usr/local/go/src/bytes/buffer.go:169 +0x1d1`

**Comment by odeke-em** [comment]
Hello there @mowenchuxin, thank you for the bug report and for the reproduction!

So:
a) So you are dereferencing a nil *os.FileInfo
<img width="758" alt="screen shot 2018-02-01 at 7 45 53 am" src="https://user-images.githubusercontent.com/4898263/35684404-0264fb60-0724-11e8-8644-9e3c16455f52.png">
where if we add a print statement
```diff
--- old.go	2018-02-01 07:49:22.000000000 -0700
+++ new.go	2018-02-01 07:49:16.000000000 -0700
@@ -215,6 +215,7 @@
 	filepath.Walk(root, func(p string, info os.FileInfo, err error) error {
 		wg.Add(1)
 		defer wg.Done()
+		fmt.Printf("p: %s\ninfo: %v\nerr: %v\n", p, info, err)
 		if info.IsDir() {
 			return nil
 		}
```
in there we'd get
```shell
p: /Users/emmanuelodeke/Desktop/openSrc/bugs/golang/23650/tool2/android_update/update_10_to_20/temp/scripts
info: <nil>
err: lstat /Users/emmanuelodeke/Desktop/openSrc/bugs/golang/23650/tool2/android_update/update_10_to_20/temp/scripts: no such file or directory
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x10a6ada]

goroutine 1 [running]:
main.md5_code.func1(0xc42001e3f0, 0x68, 0x0, 0x0, 0x10f0300, 0xc420082270, 0x0, 0x0)
	/Users/emmanuelodeke/Desktop/openSrc/bugs/golang/23650/tool2/src/md5_res/md5_res.go:219 +0x1da
path/filepath.Walk(0xc42001e3f0, 0x68, 0xc420062240, 0x0, 0x0)
	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/path/filepath/path.go:401 +0x6c
main.md5_code(0xc42001e3f0, 0x68, 0xc42001e460, 0x68, 0x8)
	/Users/emmanuelodeke/Desktop/openSrc/bugs/golang/23650/tool2/src/md5_res/md5_res.go:215 +0x14b
main.main()
	/Users/emmanuelodeke/Desktop/openSrc/bugs/golang/23650/tool2/src/md5_res/md5_res.go:35 +0x3cc
```
Please check your errors as per https://golang.org/pkg/path/filepath/#Walk
<img width="933" alt="screen shot 2018-02-01 at 7 48 26 am" src="https://user-images.githubusercontent.com/4898263/35684516-53dfcd1c-0724-11e8-9934-77844b8f13ac.png">

b) How come your utilities were built with Go1.4.2?
```shell
$ goversion tools/md5_res 
tools/md5_res go1.4.2
```
so your reported Go version should actually be go1.4.2, and I mention this because in your report, your stacktrace doesn't have a disambiguated signal number i.e it is 0xb or 11 which is SIGSEGV
```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x14 pc=0x461d]

goroutine 1 [running]:
main.func·003(0x1033c460, 0x68, 0x0, 0x0, 0x184ae8, 0x10338360, 0x0, 0x0)
```
but that's a bug that @ianlancetaylor and I fixed in 2016 for Go1.7 with commit https://github.com/golang/go/commit/1a7fc2357b1c26dcdf4fa57dee67a1172696801f aka CL https://go-review.googlesource.com/#/c/22753/. And I also noticed that the paths in the stacktrace are radically different
```diff
--- before	2018-02-01 07:58:25.000000000 -0700
+++ after	2018-02-01 07:58:12.000000000 -0700
@@ -1,12 +1,12 @@
 panic: runtime error: invalid memory address or nil pointer dereference
-[signal 0xb code=0x1 addr=0x14 pc=0x461d]
+[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x10a690a]
 
 goroutine 1 [running]:
-main.func·003(0x1033c460, 0x68, 0x0, 0x0, 0x184ae8, 0x10338360, 0x0, 0x0)
-	/Volumes/hd2/k007/tool/src/md5_res/md5_res.go:218 +0x8d
-path/filepath.Walk(0x1033c460, 0x68, 0x10354f04, 0x0, 0x0)
-	/usr/local/go/src/path/filepath/path.go:392 +0x91
-main.md5_code(0x1033c460, 0x68, 0x1033c4d0, 0x68, 0x1033c4d0)
-	/Volumes/hd2/k007/tool/src/md5_res/md5_res.go:237 +0xfb
+main.md5_code.func1(0xc42001e3f0, 0x68, 0x0, 0x0, 0x10f00e0, 0xc420082270, 0x0, 0x0)
+	/Users/emmanuelodeke/Desktop/openSrc/bugs/golang/23650/tool2/src/md5_res/md5_res.go:218 +0xba
+path/filepath.Walk(0xc42001e3f0, 0x68, 0xc420062240, 0x0, 0x0)
+	/Users/emmanuelodeke/go/src/go.googlesource.com/go/src/path/filepath/path.go:401 +0x6c
+main.md5_code(0xc42001e3f0, 0x68, 0xc42001e460, 0x68, 0x8)
+	/Users/emmanuelodeke/Desktop/openSrc/bugs/golang/23650/tool2/src/md5_res/md5_res.go:215 +0x14b
 main.main()
-	/Volumes/hd2/k007/tool/src/md5_res/md5_res.go:35 +0x2af
+	/Users/emmanuelodeke/Desktop/openSrc/bugs/golang/23650/tool2/src/md5_res/md5_res.go:35 +0x3cc
```

In summary:
a) Please check your errors in the Walk routine and avoid that nil dereference
```diff
--- old.go	2018-02-01 08:25:50.000000000 -0700
+++ new.go	2018-02-01 08:26:48.000000000 -0700
@@ -215,6 +215,10 @@
 	filepath.Walk(root, func(p string, info os.FileInfo, err error) error {
 		wg.Add(1)
 		defer wg.Done()
+		if err != nil {
+			// Check and handle the error appropriately
+			return err
+		}
 		if info.IsDir() {
 			return nil
 		}
```
b) Please try out the latest Go release Go1.10rc1 https://golang.org/dl/#unstable, 1.10 is due to be released pretty soon, 1.4.2 is no longer supported and make sure to run `make` in order to regenerate the binaries and even make them executable. If perhaps you are getting these from a vendor, let them know that their utilities could use a refresh in order to be on par with the supported. You can check what Go versions your Go binaries are built with using @rsc's https://github.com/rsc/goversion
```shell
go get rsc.io/goversion
```

I'll close this bug as the problem here is that err is not being checked yet info is being dereferenced when nil(the normal state). Please feel free to reopen if the problem is something else.

**Comment by mowenchuxin** [comment]
Thank u very much and Sorry for reply late. I am a newer to golang and this is a tool from an old program, so I do not know the version is Go1.4.2 , I fixed it with your suggest, now it can work. Thank u again.
